velox/functions/prestosql/aggregates/EntropyAggregates.cpp

```diff
@@ -0,0 +1,399 @@
+      override {
+    auto vector = (*result)->as<FlatVector<double>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+    uint64_t* rawNulls = getRawNulls(vector);
+
+    double* rawValues = vector->mutableRawValues();
+    for (int32_t i = 0; i < numGroups; ++i) {
+      char* group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        clearNull(rawNulls, i);
+        EntropyAccumulator* accData = accumulator(group);
+        rawValues[i] = accData->result();
+      }
+    }
+  }
+
+ protected:
+  inline EntropyAccumulator* accumulator(char* group) {
+    return exec::Aggregate::value<EntropyAccumulator>(group);
+  }
+
+ private:
+  // partial
+  template <bool tableHasNulls = true>
+  inline void updateNonNullValue(char* group, T value) {
+    if constexpr (tableHasNulls) {
+      exec::Aggregate::clearNull(group);
+    }
+    EntropyAccumulator* accData = accumulator(group);
+    accData->update(value);
+  }
+
+  template <bool tableHasNulls = true>
+  inline void updateNonNullValue(
+      char* group,
+      const EntropyAccumulator& accData) {
+    if constexpr (tableHasNulls) {
+      exec::Aggregate::clearNull(group);
+    }
+    EntropyAccumulator* thisAccData = accumulator(group);
+    thisAccData->merge(accData.sumC(), accData.sumCLogC());
+  }
+
+  template <bool tableHasNulls = true>
+  inline void updateNonNullValue(char* group, double sumC, double sumCLogC) {
+    if constexpr (tableHasNulls) {
+      exec::Aggregate::clearNull(group);
+    }
+    EntropyAccumulator* accData = accumulator(group);
+    accData->merge(sumC, sumCLogC);
+  }
+
+  DecodedVector decodedRaw_;
+  DecodedVector decodedPartial_;
+};
+
+// Registration code
+void checkRowType(const TypePtr& type, const std::string& errorMessage) {
+  VELOX_CHECK_EQ(type->kind(), TypeKind::ROW, "{}", errorMessage);
+  VELOX_CHECK_EQ(
+      type->childAt(kSumCIdx)->kind(), TypeKind::DOUBLE, "{}", errorMessage);
+  VELOX_CHECK_EQ(
+      type->childAt(kSumCLogCIdx)->kind(),
+      TypeKind::DOUBLE,
+      "{}",
+      errorMessage);
+}
+
+exec::AggregateRegistrationResult registerEntropy(const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures;
+  std::vector<std::string> inputTypes = {"smallint", "integer", "bigint"};
+  for (const auto& inputType : inputTypes) {
+    signatures.push_back(exec::AggregateFunctionSignatureBuilder()
+                             .returnType("double")
+                             .intermediateType("row(double,double)")
+                             .argumentType(inputType)
+                             .build());
+  }
+
+  return exec::registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType) -> std::unique_ptr<exec::Aggregate> {
```

Comment:
@liujiayi771 We need to add query config here, can you update?

Reply:
Ok, I have rebased the latest code and adapted the latest changes.